Implement "RAdamW" optimizer

bilzard opened the pull request:
Fixes #107282

## Overview

- basic design decision was followed as they made on #103881 (tensor operation, test cases, order & position of argument etc.)
- for the algorithm for decoupled weight decay, I referred to [1, 2]

## backwards-incompatible changes

- positional argument `decoupled_weight_decay` is added to:
    -  `torch.optim.radam`

The existing code which refers to these APIs can be affected.

Note: Positional argument `decoupled_weight_decay` is added to `torch.optim.RAdam`. However, since it was added to the last position and with default value, it is not affected.

## Reference

- [1] [Decoupled Weight Decay Regularization](https://arxiv.org/abs/1711.05101)
- [2] https://github.com/LiyuanLucasLiu/RAdam/blob/master/radam/radam.py#L5-L94

## TODO

- [x] implement tensor operation
- [x] implement test cases
- [x] modify doc-string
- [x] pass unit test code locally `python test/test_optim.py -k test_radam`

bilzard committed:
add decoupled_weight_decay option to RAdam

bilzard committed:
adopt lint's advice

bilzard committed:
fix docstring

bilzard committed:
fix docstring

janeyx99 commented on torch/optim/radam.py:
```diff
@@ -46,6 +57,7 @@ def __setstate__(self, state):
         for group in self.param_groups:
             group.setdefault("foreach", None)
             group.setdefault("differentiable", False)
+            group.setdefault("decoupled_weight_decay", False)
```
great catch here! i think the nadam impl forgot to do this

janeyx99 reviewed (commented):
very nice! approving CI to let things pass--could you also post a pic of the updated docs preview? One can get to the docs by going to the Python docs preview under the Helpful links in the pytorch-bot comment.

janeyx99 commented:
I looked at the preview myself: 
![image](https://github.com/pytorch/pytorch/assets/31798555/3c0f6f6a-6d6a-40b4-aa5e-c4980ec6da8e)

Looks good!

bilzard commented:
@janeyx99 Sorry, I found a mistake on the meta-code.

The first substitution of theta_t should be like:
theta_{t-1} := theta_{t-1} - …

What should I do for that?

janeyx99 commented:
I’ve canceled the merge for now but will take a look tomorrow on my laptop. if you know the right thing to do, feel free to edit the code!

bilzard committed:
fix on meta-code: should decay theta at t-1 not t

bilzard commented:
@JeanKossaifi I committed the fix.
The preview of revised doc is:
<img width="707" alt="Screenshot 2023-08-23 at 8 10 25" src="https://github.com/pytorch/pytorch/assets/36561962/a45fe8e6-7f1d-48fa-94b5-59f42d249a62">

I also uploaded [pdf version](https://github.com/pytorch/pytorch/files/12413595/pytorch-radam.pdf) of full text on RAdam documentation.
Since I am not native English speaker, I want reviewers to check if the documentation part is decent as an English sentence.

E.g. 
> This implementation provides an option to use either the original weight_decay implementation as in Adam (where the weight_decay is applied to the gradient) or the one from AdamW (where weight_decay is applied to the weight) through the decoupled_weight_decay option. When decoupled_weight_decay is set to True, it uses the AdamW style weight decay, otherwise it uses the original Adam style. This flexibility allows it to align with both the original implementation and the [author’s implementation](https://github.com/LiyuanLucasLiu/RAdam). Further information about decoupled weight decay can be found in [Decoupled Weight Decay Regularization](https://arxiv.org/abs/1711.05101).

janeyx99 commented:
@bilzard Thanks for the fix! I think it would be more appropriate to have the pattern be more like AdamW, where we have

theta_t := theta_{t-1} ... theta_{t-1}

and then the rest of the param updates use theta_t.
<img width="762" alt="image" src="https://github.com/pytorch/pytorch/assets/31798555/73149047-9c1e-465a-8118-c9a4ab1aba96">

janeyx99 commented on torch/optim/radam.py:
```diff
@@ -156,9 +172,13 @@ def step(self, closure=None):
-    is different from the `author's implementation`_.
+    This implementation provides an option to use either the original weight_decay implementation as in Adam
+    (where the weight_decay is applied to the gradient) or the one from AdamW (where weight_decay is applied
+    to the weight) through the decoupled_weight_decay option. When decoupled_weight_decay is set to True,
+    it uses the AdamW style weight decay, otherwise it uses the original Adam style. This flexibility allows
+    it to align with both the original implementation and the `author's implementation`_. Further information
```
```suggestion
    to the weight) through the decoupled_weight_decay option. When decoupled_weight_decay is set to False
    (default), it uses the original Adam style weight decay, otherwise, it uses the AdamW style which
    corresponds more closely to the `author's implementation in the RAdam paper`_. Further information
```

bilzard replied:
Good catch. The meaning becomes more clearer.

janeyx99 reviewed (commented):
Left some feedback on the phrasing

bilzard commented:
> I think it would be more appropriate to have the pattern be more like AdamW, where we have
> theta_t := theta_{t-1} ... theta_{t-1}

It's correct for AdamW because it has no if statement. However, in RAdam, we should be care both case with `decouple_weight_decay=False` and `decouple_weight_decay=True`.

Let us see. If we modify as you say,

case1) `decouple_weight_decay=False`

- (there are no first substitution)
- theta_t := theta_{t-1} ... (2nd substitution)

This pattern is current Adam style update, and the update rule is consistent. However,

case2) `decouple_weight_decay=True`

- theta_t := theta_{t-1} ... theta_{t-1} (in the 1st substitution)
- theta_t := **theta_{t-1}** ... (2nd substitution)

Then theta_{t-1}` in the second substitution is **not decayed**. This is incorrect update.

bilzard committed:
fix phrasing in AdamW docstring

Co-authored-by: Jane (Yuan) Xu <31798555+janeyx99@users.noreply.github.com>

bilzard commented:
> Shouldn't the second substitution use theta_t?

Let us see. If we write second substitution as $\theta_t := \theta_t ...$,

case1) decouple_weight_decay=False

- (there are no first substitution)
- theta_t := **theta_t** ... (2nd substitution)

This pattern is incorrect because theta_t is **not defined** at that time.

janeyx99 commented:
Oh hm yea. In the actual code it doesn't really matter because we inplace modify params as we go. I wonder if it will be better to be explicit and have theta_t := theta_{t-1}.

The reason I don't know if I want to update theta_{t-1} is because it seems to imply we're updating the previous iteration's param. hmmmmmmmm.

bilzard commented:
> In the actual code it doesn't really matter

Yes. It's just consistency.

> it seems to imply we're updating the previous iteration's param.

It seems a little odd, but I think this is just the matter of how we set bound variable.

We have second option where we introduce a new free variable tmp,
```
if decouple_weight_decay
    tmp := theta_{t-1} - lr * weight_decay * theta_{t-1} (1st substitution)
else
    tmp := theta_{t-1} (1st substitution avoiding unbound variable)

...
    theta_t := tmp ... (2nd substitution)
```

This is consistent, but I thought it's too redundant. So I chose the current update rule.

bilzard commented:
> it seems to imply we're updating the previous iteration's param.

The notion that 'theta_{t-1} represents the past' is merely a matter of human intuition. In reality, theta_{t-1} is just a variable.

janeyx99 commented:
> We have second option where we introduce a new variable tmp,

Well, we could just have tmp be theta_t, right? In the meta code?
```
if weight_decay != 0
     ...
     theta_t := theta_{t-1} - lr * weight_decay * theta_{t-1} (1st substitution)
else
     theta_t := theta_{t-1} 
...
     theta_t := theta_t ...
```

bilzard commented:
Or, the 3rd option is more preferable?
In this option, replacing tmp to theta_t in the 2nd option.

```
if decouple_weight_decay
    theta_t := theta_{t-1} - lr * weight_decay * theta_{t-1} (1st substitution)
else
    theta_t := theta_{t-1} (1st substitution avoiding unbound variable)

...
    theta_t := theta_t ... (2nd substitution)
```

janeyx99 commented:
hahahahaha nice we came to the same conclusion

bilzard commented:
Yah. I will fix to this pattern.

janeyx99 commented:
Sounds good, I will also update my NAdam PR similarly :) Thank you for your attention to detail and good catches here!

bilzard committed:
fix meta-code in RAdam

bilzard committed:
fix broken link in RAdam

bilzard commented:
@janeyx99 I fixed meta-code as to 3rd pattern.
<img width="627" alt="Screenshot 2023-08-24 at 3 30 40" src="https://github.com/pytorch/pytorch/assets/36561962/9ba5fd4d-552c-4c2c-8b06-ea3f0c4d86a2">

I also fixed [broken link](https://github.com/pytorch/pytorch/pull/107507/commits/24110c2330c8b230c32f15d211e553d35a559e4d) (which occurs on revised doc string). I update the pdf version so that you can also check the final output of the documentation.
[pytorch-optim-radam.pdf](https://github.com/pytorch/pytorch/files/12422231/pytorch-optim-radam.pdf)

janeyx99 commented:
Thanks for the quick turnaround. One more thing (hopefully the last 😛 ) is that we should just set theta_t before the if statement (otherwise if weight_decay is 0 then theta_t is again not set).

I've done something similar here https://github.com/pytorch/pytorch/pull/107736/files#diff-50e24bc32ea2cd0c850d7754c34a9617f0f3310bb6a47ecc8795133072419633R140

janeyx99 commented:
My NAdam now looks like:
<img width="569" alt="image" src="https://github.com/pytorch/pytorch/assets/31798555/ea1818cf-4c8b-4dc4-8512-3f4c6f0103e5">

bilzard commented:
> we should just set theta_t before the if statement (otherwise if weight_decay is 0 then theta_t is again not set).

Yes. Good catch. I will fix this maybe tomorrow.

As a side note, I've started to wonder if the subscript t is really necessary. Removing it and simply substituting into θ might make the whole thing simpler. However, such a change would affect for other optimizers as well, so for the scope of this discussion, the proposed change seems to be the best approach.

bilzard committed:
fix pseudocode in RAdam & fix layout

janeyx99 commented:
i did notice that, but didn’t think it was too incorrect as thankfully theta_t and theta_{t-1} are technically the same at that point. i am happy to further consider whether we want subscripts here at all—it does seem to remove lots of discrepancies, but yes, would not go in this PR.

janeyx99 commented:
feel free to merge whenever by using the pytorchbot command! let me know if it doesn’t work for you as you’re a first time contributor and i’ll help!

bilzard commented:
Hmm..., strange lint error.
I didn't edit any code but docstring, and locally running lint passed...
```
root@840554b77202:/pytorch-devenv/working/pytorch# make lint
lintrunner
⠁ FLAKE8 running...
⠁ MYPY running...
  CLANGFORMAT success!
  MYPYNOFOLLOW success!
⠁ TYPEIGNORE running...
  MYPYSTRICT success!
  CLANGTIDY success!
⠁ FLAKE8 running...
⠁ MYPY running...
  CLANGFORMAT success!
  MYPYNOFOLLOW success!
⠁ TYPEIGNORE running...
  MYPYSTRICT success!
  CLANGTIDY success!
⠁ NOQA running...
  CIRCLECI success!
⠁ FLAKE8 running...
⠁ MYPY running...
  CLANGFORMAT success!
  MYPYNOFOLLOW success!
⠁ TYPEIGNORE running...
  MYPYSTRICT success!
  CLANGTIDY success!
⠁ NOQA running...
  CIRCLECI success!
⠁ FLAKE8 running...
⠁ MYPY running...
  CLANGFORMAT success!
  MYPYNOFOLLOW success!
⠁ TYPEIGNORE running...
  MYPYSTRICT success!
  CLANGTIDY success!
⠁ NOQA running...
  CIRCLECI success!
⠁ FLAKE8 running...
⠁ MYPY running...
  CLANGFORMAT success!
  MYPYNOFOLLOW success!
⠁ TYPEIGNORE running...
  MYPYSTRICT success!
  CLANGTIDY success!
  NOQA success!
  CIRCLECI success!
⠁ FLAKE8 running...
⠁ MYPY running...
  CLANGFORMAT success!
  MYPYNOFOLLOW success!
⠁ TYPEIGNORE running...
  MYPYSTRICT success!
  CLANGTIDY success!
  NOQA success!
  CIRCLECI success!
⠁ FLAKE8 running...
⠁ MYPY running...
  CLANGFORMAT success!
  MYPYNOFOLLOW success!
⠁ TYPEIGNORE running...
  MYPYSTRICT success!
  CLANGTIDY success!
  NOQA success!
  CIRCLECI success!
⠁ FLAKE8 running...
⠁ MYPY running...
  CLANGFORMAT success!
  MYPYNOFOLLOW success!
⠁ TYPEIGNORE running...
  MYPYSTRICT success!
  CLANGTIDY success!
  NOQA success!
  CIRCLECI success!
⠁ FLAKE8 running...
⠁ MYPY running...
  CLANGFORMAT success!
  MYPYNOFOLLOW success!
⠁ TYPEIGNORE running...
  MYPYSTRICT success!
  CLANGTIDY success!
  NOQA success!
  CIRCLECI success!
⠁ FLAKE8 running...
⠁ MYPY running...
  CLANGFORMAT success!
  MYPYNOFOLLOW success!
⠁ TYPEIGNORE running...
  MYPYSTRICT success!
  CLANGTIDY success!
  NOQA success!
  CIRCLECI success!
⠁ FLAKE8 running...
⠁ MYPY running...
  CLANGFORMAT success!
  MYPYNOFOLLOW success!
  TYPEIGNORE success!
  MYPYSTRICT success!
  CLANGTIDY success!
  NOQA success!
  CIRCLECI success!
⠉ FLAKE8 running...
⠉ MYPY running...
  CLANGFORMAT success!
  MYPYNOFOLLOW success!
  TYPEIGNORE success!
  MYPYSTRICT success!
  CLANGTIDY success!
  NOQA success!
  CIRCLECI success!
⠉ FLAKE8 running...
⠙ MYPY running...
  CLANGFORMAT success!
  MYPYNOFOLLOW success!
  TYPEIGNORE success!
  MYPYSTRICT success!
  CLANGTIDY success!
  NOQA success!
  CIRCLECI success!
⠙ FLAKE8 running...
⠚ MYPY running...
  CLANGFORMAT success!
  MYPYNOFOLLOW success!
  TYPEIGNORE success!
  MYPYSTRICT success!
  CLANGTIDY success!
  NOQA success!
  CIRCLECI success!
⠚ FLAKE8 running...
⠚ MYPY running...
  CLANGFORMAT success!
  MYPYNOFOLLOW success!
  TYPEIGNORE success!
  MYPYSTRICT success!
  CLANGTIDY success!
  NOQA success!
  CIRCLECI success!
⠚ FLAKE8 running...
⠒ MYPY running...
  CLANGFORMAT success!
  MYPYNOFOLLOW success!
  TYPEIGNORE success!
  MYPYSTRICT success!
  CLANGTIDY success!
  NOQA success!
  CIRCLECI success!
⠒ FLAKE8 running...
⠂ MYPY running...
  CLANGFORMAT success!
  MYPYNOFOLLOW success!
  TYPEIGNORE success!
  MYPYSTRICT success!
  CLANGTIDY success!
  NOQA success!
  CIRCLECI success!
⠂ FLAKE8 running...
⠂ MYPY running...
  CLANGFORMAT success!
  MYPYNOFOLLOW success!
  TYPEIGNORE success!
  MYPYSTRICT success!
  CLANGTIDY success!
  NOQA success!
  CIRCLECI success!
  FLAKE8 success!
⠂ MYPY running...
  CLANGFORMAT success!
  MYPYNOFOLLOW success!
  TYPEIGNORE success!
  MYPYSTRICT success!
  CLANGTIDY success!
  NOQA success!
  CIRCLECI success!
  FLAKE8 success!
⠒ MYPY running...
  CLANGFORMAT success!
  MYPYNOFOLLOW success!
  TYPEIGNORE success!
  MYPYSTRICT success!
  CLANGTIDY success!
  NOQA success!
  CIRCLECI success!
  FLAKE8 success!
⠲ MYPY running...
  CLANGFORMAT success!
  MYPYNOFOLLOW success!
  TYPEIGNORE success!
  MYPYSTRICT success!
  CLANGTIDY success!
  NOQA success!
  CIRCLECI success!
  FLAKE8 success!
⠴ MYPY running...
  CLANGFORMAT success!
  MYPYNOFOLLOW success!
  TYPEIGNORE success!
  MYPYSTRICT success!
  CLANGTIDY success!
  NOQA success!
  CIRCLECI success!
  FLAKE8 success!
⠤ MYPY running...
  CLANGFORMAT success!
  MYPYNOFOLLOW success!
  TYPEIGNORE success!
  MYPYSTRICT success!
  CLANGTIDY success!
  NOQA success!
  CIRCLECI success!
  FLAKE8 success!
⠤ MYPY running...
  CLANGFORMAT success!
  MYPYNOFOLLOW success!
  TYPEIGNORE success!
  MYPYSTRICT success!
  CLANGTIDY success!
  NOQA success!
  CIRCLECI success!
  FLAKE8 success!
  MYPY success!
  CLANGFORMAT success!
  MYPYNOFOLLOW success!
  TYPEIGNORE success!
  MYPYSTRICT success!
  CLANGTIDY success!
  NOQA success!
  CIRCLECI success!
  NATIVEFUNCTIONS success!
  NEWLINE success!
  SPACES success!
  TABS success!
  CONSTEXPR success!
  INCLUDE success!
  PYBIND11_INCLUDE success!
  ERROR_PRONE_ISINSTANCE success!
  EXEC success!
  PYPIDEP success!
  PYBIND11_SPECIALIZATION success!
  CUBINCLUDE success!
  RAWCUDA success!
  RAWCUDADEVICE success!
  ROOT_LOGGING success!
  DEPLOY_DETECTION success!
  SHELLCHECK success!
  CMAKE success!
  ACTIONLINT success!
  TESTOWNERS success!
  CALL_ONCE success!
  ONCE_FLAG success!
  WORKFLOWSYNC success!
  UFMT success!
  COPYRIGHT success!
  BAZEL_LINTER success!
  RUFF success!
  LINTRUNNER_VERSION success!
ok No lint issues.
```

https://github.com/pytorch/pytorch/actions/runs/5958614027/job/16163595945?pr=107507#step:18:57

bilzard commented:
> i did notice that, but didn’t think it was too incorrect as thankfully theta_t and theta_{t-1} are technically the same at that point.

Indeed, that's a valid point. However, caution is needed if we plan to update $\theta_t$ before the next reference of $\theta_{t-1}$ in the future. Personally, I think it's safer to always refer to the latest $\theta_t$, but it does seem like consistency is maintained as it stands.

Another option is using no-subscripted $\theta$ in the middle of the algorithm, and finally update $\theta_t$ in the last. 
But I think its optional to this PR or NAdam's PR.

```
theta := theta_{t-1}
...
(constantly update and refer to theta in the middle of algorithm)
...
theta_t := theta ... (finally update theta_t)
```

By the way, could you have any idea of above linting error I shared?

janeyx99 commented:
Yea, we can open an issue to track the theta subscripting clarification if you would like, but it is not very high priority. The linting CI failure is definitely not your PR's fault but seems to be some infrastructure problem with GitHub Actions. Let's wait for the rest of the signal and rerun some of them if everything else is green.

bilzard commented:
> Let's wait for the rest of the signal and rerun some of them if everything else is green.

@janeyx99 
I see. Since I can't distinguish the error is from this PR and other factor, could you cast merge request to torchbot if the test ended?

bilzard commented:
@janeyx99 I added "backwards-incompatible changes" section in the description of PR. Could you also check this?

janeyx99 commented:
Yea, weird. I'm gonna submit a merge request with a rebase.

@pytorchbot merge -r

janeyx99 commented:
@ZainRizvi do you have any ideas why the CI is failing like this?

ZainRizvi commented:
Looks like a flaky issue that succeeded on retry

janeyx99 commented:
@pytorchbot merge -r
